plugins/kubernetes/app/models/watchers/deploy_watcher.rb

```diff
@@ -1,119 +1,254 @@
+        send_event(release_doc)
+        terminate_watcher if deploy_finished?(project)
+      end
     end
 
-    def update_replica_count(release_doc, pod_event)
-      rc = rc_pods(release_doc.replication_controller_name)
+    def handle_pod_update(pod)
+      update_rc_pods(pod)
+      release = Kubernetes::Release.find(pod.release_id)
+      release_doc = release.release_doc_for(pod.deploy_group_id, pod.role_id)
+
+      live_pods = count_live_pods(pod.rc_unique_identifier)
+      failed_pods = count_failed_pods(pod.rc_unique_identifier)
 
-      if pod_event.deleted?
-        rc.delete(pod_event.pod.name)
+      if !recoverable_fail?(release_doc, failed_pods)
+        update_failed_release_doc(live_pods, release_doc) { yield release_doc if block_given? }
       else
-        rc[pod_event.pod.name] = pod_event.pod
+        update_release_doc(release_doc, live_pods) { yield release_doc if block_given? }
       end
+    end
+
+    def recoverable_fail?(release_doc, failed_pods)
+      release_doc.failed? && release_doc.can_recover?(failed_pods)
+    end
+
+    def update_release_doc(release_doc, live_pods)
+      reset_timeout
 
-      ready_count = rc.reduce(0) { |count, (_pod_name, pod)| count += 1 if pod.ready?; count }
-      release_doc.update_replica_count(ready_count) unless release_doc.replicas_live == ready_count
-      @release.update_columns(status: :spinning_up) if @release.created?
+      if release_doc.live_replicas_changed?(live_pods)
+        release_doc.update_status(live_pods)
+        release_doc.update_replica_count(live_pods)
+        release_doc.update_release
+        yield
+      end
     end
 
-    def update_timeout(pod_event)
-      unless pod_event.deleted?
-        @pod_timer.reset if pod_event.pod.ready?
+    def update_failed_release_doc(live_pods, release_doc)
+      if release_doc.live_replicas_changed?(live_pods)
+        release_doc.update_replica_count(live_pods)
+        yield
       end
     end
 
-    def end_deploy
-      @release.release_is_live!
-      info 'Deploy is live!'
-      terminate
+    # Involved object identifies the Kubernetes resource that triggered the event
+    def involved_object_for(notice)
+      notice.object.involvedObject
+    end
+
+    def cluster_for(cluster_id)
+      Kubernetes::Cluster.find(cluster_id)
+    end
+
+    def rcs
+      @rcs ||= {}
+    end
+
+    def rc_pods(rc_unique_id)
+      rcs[rc_unique_id] ||= {}
+      rcs[rc_unique_id]
+    end
+
+    def update_rc_pods(pod)
+      rc = rc_pods(pod.rc_unique_identifier)
+      rc[pod.name] = pod
+    end
+
+    def release_ids_from_cluster
+      rcs.map { |_, pods| pods.map { |_, pod| pod.release_id } }.flatten.uniq
+    end
+
+    def last_release(project)
+      project.kubernetes_releases.last
+    end
+
+    def count_live_pods(rc_unique_identifier)
+      rc = rc_pods(rc_unique_identifier)
+      rc.reduce(0) { |count, (_, pod)| count += 1 if pod.live? && !deleted?(pod); count }
```

Comment:
Humm, that's a good suggestion, yes. That hasn't crossed my mind because I've been treating the `live?` and `deleted?` concepts as completely independent, as this information comes from different sources, but no harm should come from doing that. I'll give it a try.